crates/router/src/core/locking.rs

```diff
@@ -0,0 +1,191 @@
+
+    
+    if is_locking_enabled_for_merchant && is_locking_enabled_on_api {
+        let expiry_in_seconds = 60; // get it from redis
+        let delay_between_retries_in_seconds = 10; // get it from redis
+        let retries = 1; // get from redis based on should_retry, if not present in redis default 1?
+        let locking_key = locking_input.unique_locking_key;
+        lock_resource(
+            state,
+            locking_key,
+            expiry_in_seconds,
+            retries,
+            delay_between_retries_in_seconds,
+            api_identifier,
+        )
+        .await
+    } else {
+        logger::info!(
+            "Resource Locking not Enabled for merchant_id: {} and api: {}",
+            locking_input.merchant_id,
+            api_identifier.to_owned()
+        );
+        Ok(LockStatus::NotEnabled)
+    }
+}
+
+#[instrument(skip(state))]
+pub async fn lock_resource(
+    state: &routes::AppState,
+    locking_key: String,
+    expiry_in_seconds: u64,
+    retries: i32,
+    delay_between_retries_in_seconds: i64,
+    _api_identifier: &str,
+) -> RouterResult<LockStatus> {
+    let redis_key_for_lock = get_redis_key_for_locks(locking_key);
+
+    acquire_lock_on_resource_in_redis(
+        state,
+        redis_key_for_lock.as_str(),
+        true,
+        expiry_in_seconds,
+        delay_between_retries_in_seconds,
+        retries,
+    )
+    .await
+}
+
+#[inline]
+fn get_redis_key_for_locks(a: String) -> String {
+    format!("API_LOCK_{}", a)
+}
+
+#[instrument(skip(state))]
+pub async fn acquire_lock_on_resource_in_redis(
+    state: &routes::AppState,
+    key: &str,
+    value: bool,
+    expiry_in_seconds: u64,
+    _delay_between_retries_in_seconds: i64,
+    mut retries: i32,
+) -> RouterResult<LockStatus> {
+    let redis_conn = state.store.clone().get_redis_conn();
+
+    while retries != 0 {
+        retries -= 1;
+        let is_lock_acquired = redis_conn
+            .set_key_if_not_exists_with_expiry(
+                key,
+                value,
+                Some(
+                    expiry_in_seconds
+                        .try_into()
+                        .into_report()
+                        .change_context(errors::ApiErrorResponse::InternalServerError)?,
+                ),
+            )
+            .await;
+
+        match is_lock_acquired {
+            Ok(redis::SetnxReply::KeySet) => {
+                return Ok(LockStatus::Acquired(key.to_owned()));
+            }
+            Ok(redis::SetnxReply::KeyNotSet) => {
+                logger::error!("Lock not acquired, retrying");
+                actix_time::sleep(tokio::time::Duration::from_millis(expiry_in_seconds * 1000))
+                    .await;
+            }
+            Err(error) => {
+                logger::error!(error=%error.current_context(), "Error while locking");
+                actix_time::sleep(tokio::time::Duration::from_millis(expiry_in_seconds * 1000))
+                    .await;
+            }
```

Comment:
Can we have like a seperate interface for retry like
```rust
struct Retry;
impl Retry {
	fn perform(fun: F,retries: i32) 
	where
		F: FnOnce-> Result<T,E>
	{
	}
}
```
Or a macro since this could be used in multiple places

Reply:
This will be used only here. No multiple place implementation.

Reply:
Retry is a generic mechanism you can create an interface and in future if you wanna use it somewhere else you can use it